### ISSUE TITLE
Reuse models from models

### ISSUE THREAD

**Issue by tzachshabtay** [go-swagger/go-swagger#1912]
Our use-case is that we have `common.yml` with common type definitions, and we have `service1.yml` and `service2.yml` with refs to `common.yml`, i.e:

```
...
schema:
  $ref: '../common.yml#/definitions/Status'
```

If we do `swagger generate model -f service1.yml` and `swagger generate model -f service2.yml` the definitions in `common.yml` are duplicated in both services (we generate the models for both services to different folders).
We would like a way to generate models for `common.yml` separately and then specify it when generating our services (similarly to how we use `existing-models` when generating the server and the client), for example:
`swagger generate model -f service1.yml --existing-models=../common`

The documentation for `generate models` [seems to imply that such a flag actually exists](https://github.com/go-swagger/go-swagger/blob/master/docs/generate/model.md), btw, but it seems to be [ignored in code](https://github.com/go-swagger/go-swagger/blob/deefce1ed5785664f09b259c444c4142020925e1/cmd/swagger/commands/generate/model.go#L39)?

An alternative solution (less preferable but maybe workable), is to allow generating model files while keeping the hierarchy of the original spec files, i.e if my folder hierarchy looks like this:

```
| common.yml
| service1
| - service1.yml
| service2
| - service2.yml
```

Then generating the models for service 1 will give me:
```
| status.go
| service1
| - service1_status.go
```

Is one of this options already possible (I didn't see any way)?

Thanks.

**Comment by fredbi** [comment]
I one of your spec, you can specify you want to reuse with a `x-go-type` extension.

**Comment by fredbi** [comment]
Could you please details your command line when trying to use `--existing-models` . This is supposed to work, but maybe for some reason, the generator can't resolve the proper import?

**Comment by tzachshabtay** [comment]
My command:
```
swagger generate model --existing-models=svc/api/models -f ${PROTOCOL_HOME}/api/$(service)/$(service).yml --template-dir ${TEMPLATES_HOME}/swagger/templates .
```

I'll try using `x-go-type` and report back, thanks.

**Comment by fredbi** [comment]
I believe that the package path for `--existing-models` must be a fully qualified go import path, 
like `github.com/tzachshabtay/models` .

**Comment by tzachshabtay** [comment]
Maybe I'm misunderstanding how imports work in golang, but `svc/api/models` is the same path I use to import my 'common' models in my go code (successfully). So doesn't this make it 'fully qualified'?

And repeating my question from the first post, how can this work when it's [ignored in the code](https://github.com/go-swagger/go-swagger/blob/deefce1ed5785664f09b259c444c4142020925e1/cmd/swagger/commands/generate/model.go#L39)?

About using `x-go-type`: from what I can see in this [example](https://github.com/go-swagger/go-swagger/blob/312366608bbf17dd219190b66ab63bdc8b4d0db0/fixtures/codegen/existing-model.yml#L99-L103) (and I didn't find any api documentation on it besides a small excerpt pointing to the example) you're supposed to put `x-go-type` next to the type you want to skip generating for. But doing that means I'll need to put it on my types from `common.yml`, so just generating the common models separately won't be possible (and so there will be nothing to point the service to). I tried putting this in the `ref` section in my service but it seemed to be ignored (i.e something like this):

```
paths:
  /status:
    get:
      security:
        - key: []
      tags:
        - status
      summary: Get Status
      description: Get status of service
      operationId: get#status
      responses:
        '200':
          description: OK
          schema:
            $ref: '../common.yml#/definitions/Status'
            x-go-type:
              import:
                package: "svc/api/models"
                alias: "status"
              type: "Key"
```

**Comment by gregmarr** [comment]
I don't think that's supposed to be used in `generate model`, just `generate client` and `generate server`.

https://github.com/go-swagger/go-swagger/commit/a3769346e67303c88f698713045981e709a45f42

The docs were written later.

**Comment by fredbi** [comment]
You are right, anything next to a $ref is ignored (this comes from jsonschema spec afaik)

**Comment by tzachshabtay** [comment]
We found a workaround that seems to work well:
- We define our type `Status` (for example) in `common.yml`.
- Then, in each of our services that need `Status` we'll wrap it in another `Status` type and add both `allOf` (that refers to the "real" status type) and `x-go-type`, for example:

```
  Status:
    allOf:
      - $ref: '../common.yml#/definitions/Status'
    x-go-type:
      import:
        package: "svc/api/models"
        alias: "common"
      type: "Status"
```

Now we can generate our common models separately, and generating our service models will create the expected imports and skip generating code.

It will still be nice to have this feature supported so we won't need this extra boilerplate in the yml files, but at least we have something that works and allows us to share our models between different services.
